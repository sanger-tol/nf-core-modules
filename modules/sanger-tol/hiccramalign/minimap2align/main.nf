process HICCRAMALIGN_MINIMAP2ALIGN {
    tag "$meta.id"
    label "process_high"

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://community-cr-prod.seqera.io/docker/registry/v2/blobs/sha256/65/65858e733832166824cfd05291fc456bdf219b02baa3944c2c92efad86a6ee7f/data' :
        'community.wave.seqera.io/library/htslib_minimap2_samtools_gawk_perl:6729620c63652154' }"

    input:
    tuple val(meta), path(cram), path(crai), val(chunkn), val(range), path(reference)

    output:
    tuple val(meta), path("*.bam"), emit: bam
    path "versions.yml"           , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    // WARNING: This module includes the filter_five_end.pl script as a module binary in
    // ${moduleDir}/resources/usr/bin/filter_five_end.pl. To use this module, you will
    // either have to copy this file to ${projectDir}/bin or set the option
    // nextflow.enable.moduleBinaries = true
    // in your nextflow.config file.
    def args1 = task.ext.args1 ?: ''
    def args2 = task.ext.args2 ?: ''
    def args3 = task.ext.args3 ?: ''
    def args4 = task.ext.args4 ?: ''
    def args5 = task.ext.args5 ?: ''
    def args6 = task.ext.args6 ?: ''
    def prefix  = task.ext.prefix ?: "${cram}.${chunkn}.${meta.id}"
    """
    samtools cat ${args1} -r "#:${range[0]}-${range[1]}" ${cram} |\\
        samtools fastq ${args2} - |\\
        minimap2 -t${task.cpus} ${args3} ${reference} - |\\
        gawk '
            \$1 ~ /^\\@/ { print \$0 }
            \$1 !~ /^\\@/ && and(\$2, 64) > 0 { print 1 \$0 }
            \$1 !~ /^\\@/ && and(\$2, 64) == 0 { print 2 \$0 }
        ' |\\
        filter_five_end.pl |\\
        gawk '
            BEGIN { OFS="\\t" }
            \$1 ~ /^\\@/ { print \$0 }
            \$1 !~ /^\\@/ { \$2 = and(\$2, compl(2048)); print substr(\$0, 2) }
        ' |\\
        samtools fixmate ${args4} - - |\\
        samtools view -h ${args5} |\\
        samtools sort ${args6} -@${task.cpus} -T ${prefix}_tmp -o ${prefix}.bam -

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        samtools: \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//' )
        minimap2: \$(minimap2 --version | sed 's/minimap2 //g')
        gawk: \$(gawk --version | grep -o -E "[0-9]+(\\.[0-9]+)+" | head -n1)
    END_VERSIONS
    """

    stub:
    def prefix  = task.ext.prefix ?: "${cram}.${chunkn}.${meta.id}"
    """
    touch ${prefix}.bam

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        samtools: \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//' )
        minimap2: \$(minimap2 --version | sed 's/minimap2 //g')
        gawk: \$(gawk --version | grep -o -E "[0-9]+(\\.[0-9]+)+" | head -n1)
    END_VERSIONS
    """
}
