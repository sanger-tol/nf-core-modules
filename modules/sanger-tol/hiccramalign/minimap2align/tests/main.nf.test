
nextflow_process {

    name "Test Process HICCRAMALIGN_MINIMAP2ALIGN"
    script "../main.nf"
    process "HICCRAMALIGN_MINIMAP2ALIGN"

    tag "modules"
    tag "modules_sangertol"
    tag "minimap2/index"
    tag "hiccramalign"
    tag "hiccramalign/minimap2align"
    tag "nf-core/minimap2/index"

    setup {
        nfcoreSetup(launchDir)
        nfcoreInstall(launchDir, ["minimap2/index"])
        nfcoreLink(launchDir, baseDir)

        run("MINIMAP2_INDEX") {
            script "../../../../nf-core/minimap2/index/main.nf"
            process {
                """
                input[0] = Channel.of([
                    [ id: "test" ],
                    file(params.modules_testdata_base_path + 'Meles_meles/assembly/release/mMelMel3.1_paternal_haplotype/GCA_922984935.2.subset.fasta.gz', checkIfExists: true)
                ])
                """
            }
        }
    }

    test("meles meles") {
        config "./nextflow.config"

        when {

            params {
                samtools_cat_args     = ""
                samtools_fastq_args   = "-F0xB00 -nt"
                minimap2_args         = "-ax sr --no-pairing"
                samtools_fixmate_args = "-mpu"
                samtools_view_args    = "-q 0 -F 0x904"
                samtools_sort_args    = "--write-index -l1"
            }

            process {
                """
                input[0] = Channel.of([
                    [ id:'test' ], // meta map
                    file(params.modules_testdata_base_path + 'Meles_meles/genomic_data/mMelMel3/hic-arima2/35528_2%231_subset.cram', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'Meles_meles/genomic_data/mMelMel3/hic-arima2/35528_2%231_subset.cram.crai', checkIfExists: true),
                    1,
                    [0, 1]
                ]).join(MINIMAP2_INDEX.out.index)
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

        cleanup {
            nfcoreCleanup(launchDir, baseDir)
        }

    }

    test("meles meles - stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = Channel.of([
                    [ id:'test' ], // meta map
                    file(params.modules_testdata_base_path + 'Meles_meles/genomic_data/mMelMel3/hic-arima2/35528_2%231_subset.cram', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'Meles_meles/genomic_data/mMelMel3/hic-arima2/35528_2%231_subset.cram.crai', checkIfExists: true),
                    1,
                    [0, 1]
                ]).join(MINIMAP2_INDEX.out.index)
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

        cleanup {
            nfcoreCleanup(launchDir, baseDir)
        }

    }
}
