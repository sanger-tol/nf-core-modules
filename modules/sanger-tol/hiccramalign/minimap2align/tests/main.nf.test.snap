{
    "meles meles": {
        "content": [
            "3196a1997c55d543d773a1af0595fa6d",
            [
                "versions.yml:md5,92d058f31f8777917a6a79ba98abb3fc"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.3"
        },
        "timestamp": "2025-07-16T16:35:19.628869"
    },
    "meles meles - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "35528_2%231_subset.cram.1.test.bam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,92d058f31f8777917a6a79ba98abb3fc"
                ],
                "bam": [
                    [
                        {
                            "id": "test"
                        },
                        "35528_2%231_subset.cram.1.test.bam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,92d058f31f8777917a6a79ba98abb3fc"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.3"
        },
        "timestamp": "2025-07-16T16:31:49.320246"
    }
}